active-rfcs/0000-recursive-components.md

```diff
@@ -0,0 +1,150 @@
+-   Start Date: Date: 2021-12-15
+-   Reference Issues: https://github.com/withastro/rfcs/discussions/45
+-   Implementation PR:
+
+# Summary
+
+Allow a way for astro components to render themselves recursively.
+
+Nested data structures are everywhere. Common examples include nested blog comments,
+file explorer trees, extensive navigation menu trees, or using a headless CMS and nested content blocks to render an entire website.
+
+Even when the entire tree structure is KNOWN, for anything more than a few levels deep,
+using a recursive Component/function for rendering is the most efficient and elegant approach.
+
+When the tree structure is UNKNOWN, and you want to support N levels deep, using a recursive
+Component/function is the ONLY approach that will work. This is often the case
+when fetching data from an API.
+
+Svelte, which also uses the same "single component per file" pattern has already
+encountered and solved for this issue by exposing the [svelte:self](https://svelte.dev/docs#svelte_self)
+attribute as part of their API.
+
+# Example
+
+We have a set of items we want to render, stored in an object, indexed by id.
+Each item optionally defines an array of childIds.
+
+```
+const items = {
+    "1": {
+        "id": "1",
+        "childIds": ["1-1","1-2"]
+    },
+    "1-1": {
+        id: "1-1",
+        childIds: ["1-1-1", "1-1-2"]
+    },
+    "1-2": {
+        id: "1-2",
+        childIds: []
+    },
+    "1-1-1": {
+        id: "1-1-1",
+        childIds: []
+    },
+    "1-1-2": {
+        id: "1-1-2",
+        childIds: []
+    },
+    "1-1-2-1": {
+        id: "1-1-2-1",
+        childIds: []
+    },
+    "1-1-2-2": {
+        id: "1-1-2-2",
+        childIds: []
+    },
+};
+```
+
+We want to use a recursive component to render the entire tree,
+(or one of its branches) by passing in both the entire tree (the items object)
+and an itemId like this:
+
+```
+// src/components/RecursiveItem.astro
+---
+const {itemId, items} = Astro.props;
+const {childIds} = items[itemId];
+const Self = Astro.self;
+---
+<li>
+    Item: {itemId}
+    {childIds.length ? (
+        <ol>
+            {childIds.map(childId => (
+                <Astro.self itemId={childId} items={items} />
+            ))}
+        </ol>
+    ) : ""}
+</>
+```
+
+Note that in the above example, the `<Self />` component provides
+access to this component's render function, allowing it to
+reference itself.
+
+# Motivation
+
+Handling this use case lets .astro components do things that other component
+frameworks can do, making it a 1st class component framework in its own right.
```

Comment:
As said above, moving some of the paragraphs from the Summary to here would help this section.

Also, I would write up an example of how you have to do this today (probably by adding a second component).